### PR TITLE
Purge Newer Applications Deployed on Rollback

### PR DESCRIPTION
**Pull request by jonathan-innis**
- We still need to add purging of dependent applications as it's possible that there are dependencies between the newer applications that are rolled out.
- Let's discuss how we can handle this scenario

### CODE REVIEW

File: pkg/utils/helpers.go [hunk 1]
```diff
@@ -46,6 +46,15 @@ func TruncateString(in string, num int) string {
 	return out
 }
 
+func Remove(s []string, r string) []string {
+	for i, v := range s {
+		if v == r {
+			return append(s[:i], s[i+1:]...)
```
Comment (nitishm) [inline pkg/utils/helpers.go:52]: if `i == len(s)` i.e. the last element this would panic with an out of bounds

**Review by nitishm — CHANGES_REQUESTED** [review]
Check for out-of-bounds if when removing last application in the `Remove()` function

**Comment by nitishm** [comment]
Looks good overall - maybe a few test cases for the `Remove()` function or rather the `getDiff()` function would be good.

**Comment by jonathan-innis** [comment]
I'm slightly concerned about this in the context of HelmReleases with dependencies. If we were to purge HelmReleases due to a rollback for things that have dependencies between each other, is it possible that we may end up in a state where something can't be deleted? Would we have to build a reverse dependency graph for this as well? @nitishm any thoughts on this?

File: pkg/utils/helpers.go (hunk 1 above)
Comment (jonathan-innis) [reply pkg/utils/helpers.go:52]: That's not true, Because `i + 1 = len(s)` which means that `s[i+1:] = s[len(s):] = []string{}`

Comment (jonathan-innis) [reply pkg/utils/helpers.go:52]: Still added the test cases because we definitely need the code coverage 😃

**Comment by nitishm** [comment]
> I'm slightly concerned about this in the context of HelmReleases with dependencies. If we were to purge HelmReleases due to a rollback for things that have dependencies between each other, is it possible that we may end up in a state where something can't be deleted? Would we have to build a reverse dependency graph for this as well? @nitishm any thoughts on this?

Can you walk me through a scenario? Do you mean deleting a DAG node with child node(s) and parent node(s)?

**Comment by jonathan-innis** [comment]
> > I'm slightly concerned about this in the context of HelmReleases with dependencies. If we were to purge HelmReleases due to a rollback for things that have dependencies between each other, is it possible that we may end up in a state where something can't be deleted? Would we have to build a reverse dependency graph for this as well? @nitishm any thoughts on this?
> 
> Can you walk me through a scenario? Do you mean deleting a DAG node with child node(s) and parent node(s)?

Yes, as in I deploy applications through AppGroup with the following:
```
Applications
-----
A
```
Then we deploy the following set of applications in generation 2 with the following dependency structure
```
Applications
-----
A -> B -> C -> D
```
This rollout fails on D so we purge everything. However, would it not be possible that C needs B to be present for removal so this purging would fail?

**Comment by nitishm** [comment]
Yep it seems like removal should use the reversal DAG as well.

**Comment by nitishm** [comment]
app-removal-rollback

Removal Scenarios:
1. Root Node,
2. Leaf Node,
3. Middle Node
4. DAG (leaf)

Solution:
- Build reverse DAG
- Mark applications to be removed in reverse DAG
- Delete *marked* applications only following reverse DAG
- Reapply last successful spec

**Comment by jonathan-innis** [comment]
> app-removal-rollback
> 
> Removal Scenarios:
> 
> 1. Root Node,
> 2. Leaf Node,
> 3. Middle Node
> 4. DAG (leaf)
> 
> Solution:
> 
> * Build reverse DAG
> * Mark applications to be removed in reverse DAG
> * Delete _marked_ applications only following reverse DAG
> * Reapply last successful spec

Yep, this is what I was thinking about 😃

**Comment by nitishm** [comment]
> > app-removal-rollback
> > Removal Scenarios:
> > 
> > 1. Root Node,
> > 2. Leaf Node,
> > 3. Middle Node
> > 4. DAG (leaf)
> > 
> > Solution:
> > 
> > * Build reverse DAG
> > * Mark applications to be removed in reverse DAG
> > * Delete _marked_ applications only following reverse DAG
> > * Reapply last successful spec
> 
> Yep, this is what I was thinking about 😃

Let's try to enhance the existing **delete-all** using reversal to support this case (i.e delete-all means mark all nodes)

**Review by nitishm — APPROVED** [review]
🚀 Brilliant stuff

**Pull request merged**